### PR TITLE
drivers: nrf: Add GPIO shim using NRFX

### LINKED ISSUE #8713: add DTS gpio support for NRF51
The NRF51 platforms / SoCs need dts support for HAS_DTS_GPIO

### PR DESCRIPTION
**Pull request by Mierunski**
This PR adds shim for GPIO based on NRFX drivers.

It also adds DTS support for GPIO on nRF51 

Fixes #8713

### CODE REVIEW

File: include/misc/util.h
Comment (carlescufi) [inline include/misc/util.h]: Why the leading underscores? `_var`, `_bit`, `_val`

**Review by carlescufi — COMMENTED** [review]
Just a small comment but overall looks fine.

File: dts/arm/nordic/nrf51822.dtsi
Comment (ioannisg) [inline dts/arm/nordic/nrf51822.dtsi]: Something looks wrong, here: why do we have two lines for region range? This should be for nRF52840 that has P0 and P1.

**Review by ioannisg — CHANGES_REQUESTED** [review]

File: dts/arm/nordic/nrf51822.dtsi
Comment (Mierunski) [reply dts/arm/nordic/nrf51822.dtsi]: I created it based on dts for nRF52832, I will change it to one region

**Comment by ioannisg** [comment]
One question about HAS_DTS_GPIO:
- is it supposed to be added in all nRF51-based boards, or not?
- if yes, could we, instead, add it in the driver Kconfig file, as we do it for nRF52?

File: drivers/gpio/gpio_nrfx.c
Comment (ioannisg) [inline drivers/gpio/gpio_nrfx.c]: Here you are still using the "NRF5" nomenclature, which should not be done for the nrfx driver.
I understand that this might need some more work on the fixup/dts.

**Review by ioannisg — CHANGES_REQUESTED** [review]

**Comment by ioannisg** [comment]
Also, GPIO_NRF5_xxx is still used in several places; IMO it would be good idea to generate GPIO_NRFX_xxx information through DTS and FIXUP files

File: drivers/led_strip/Kconfig.ws2812b_sw
Comment (ioannisg) [inline drivers/led_strip/Kconfig.ws2812b_sw]: Could we not use a hard-coded value, here?

**Comment by Mierunski** [comment]
@ioannisg I've added HAS_DTS_GPIO to driver Kconfig file, as now all nRF boards have dts information about gpio

File: drivers/led_strip/Kconfig.ws2812b_sw
Comment (Mierunski) [reply drivers/led_strip/Kconfig.ws2812b_sw]: It was set before to port name from Kconfig file, currently Kconfig option was removed in favor of dts configuration, there is no way to reference a dts symbol in Kconfig thats why I used hardcoded value, other option is to remove this line completely

File: boards/arm/nrf51_blenano/Kconfig.board
Comment (galak) [inline boards/arm/nrf51_blenano/Kconfig.board]: Why here and not in the drivers/gpio Kconfig?

**Review by galak — DISMISSED** [review]
I don't think you need to select HAS_DTS_GPIO in boards/arm/*/Kconfig.board, since we have that in drivers/gpio/Kconfig.nrfx

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: Perhaps `inverted` would be less ambiguous.

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: Same story as in `gpio_nrfx_config` (pin will be 0).

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: This line is the only difference between `gpio_nrfx_enable_callback` and `gpio_nrfx_disable_callback`. You should consider merging these two functions.

**Review by anangl — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: I'm afraid this will be incomprehensible for the majority of sane people. Add at least a word of comment, please.

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: Sorry, I don't get it. Why not clear the PORT event? It will generate the IRQ again, won't it?

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: This gives a warning when `GPIO_COUNT` is 1.

**Review by anangl — CHANGES_REQUESTED** [review]

File: boards/arm/nrf51_blenano/Kconfig.board
Comment (Mierunski) [reply boards/arm/nrf51_blenano/Kconfig.board]: I removed it and left it only in drivers/gpio/Kconfig.nrfx

**Comment by Mierunski** [comment]
@anangl All requested changes should be addressed now

File: drivers/gpio/Kconfig.nrfx
Comment (Mierunski) [reply drivers/gpio/Kconfig.nrfx]: Changed name to "nRF GPIO nrfx driver"

**Review by anangl — CHANGES_REQUESTED** [review]

**Comment by Mierunski** [comment]
@anangl All done

**Comment by Mierunski** [comment]
@ioannisg @galak Could you revisit?

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: I suggest to rephrase it as follows:
The PORT event is generated on the rising edge of the DETECT signal, i.e. when a pin state changes to the level configured to be sensed. If any level triggering pins are still in their active state, don't clear the PORT event so that the interrupt is fired again and the proper callbacks are executed.

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: This should be zeroed here. And the calls to `check_level_trigger_pins` below should be done only when the PORT event is set as there is no need to call them otherwise. Besides, this way you hide some logic here and it may generate confusion (and errors) when some modifications are needed here in some future.

And what about renaming it to `fired_triggers`?

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: There is no need to update `int_trigger_pins` here. Just do:
```
	active_level_triggers =
#ifdef CONFIG_GPIO_NRFX_P0
		check_level_trigger_pins(DEVICE_GET(gpio_nrfx_p0)) |
#endif
#ifdef CONFIG_GPIO_NRFX_P1
		check_level_trigger_pins(DEVICE_GET(gpio_nrfx_p1)) |
#endif
		0;
```
And this should be done only when the PORT event is set. Then also the `active_level_triggers` can be local to the related `if` block.

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: `u8_t`, to stay consistent.

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: `u8_t`, to stay consistent.

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: Such indentation is inconsistent with what you've got right below. IMO it would be better to write it this way:
```
		nrf_gpio_cfg(NRF_GPIO_PIN_MAP(get_port_cfg(port)->port_num,
					      curr_pin),
			     dir, input, pull, drive, NRF_GPIO_PIN_NOSENSE);
```

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: Why is it that complicated?
The loop should be `for (u8_t channel = 0; channel < GPIOTE_CH_NUM; ++channel)` and when an unused channel is found it should be configured and the function should return its index. Then after the loop only `return -ENODEV;` is sufficient.

File: drivers/gpio/gpio_nrfx.c
Comment (anangl) [inline drivers/gpio/gpio_nrfx.c]: Such check should be added to the GPIOTE HAL. Add a TODO comment for this, please.

**Review by anangl — CHANGES_REQUESTED** [review]

**Review by anangl — COMMENTED** [review]
LGTM. Some final polishing suggested.

File: drivers/gpio/gpio_nrfx.c
Comment (Mierunski) [reply drivers/gpio/gpio_nrfx.c]: I think that it will needlessly complicate code at line 378, while saving only 4 bytes of ram. I would like too keep it in current way.

Comment (anangl) [reply drivers/gpio/gpio_nrfx.c]: Right. I somehow forgot that this array was used also there. Sure, keep it as it is.

**Review by anangl — APPROVED** [review]

File: dts/bindings/gpio/nordic,nrf-gpio.yaml
Comment (anangl) [inline dts/bindings/gpio/nordic,nrf-gpio.yaml]: NRF5 -> nRF
Same applies to [line 12](https://github.com/zephyrproject-rtos/zephyr/pull/8729/files#diff-75b66b32098c53e8beb8efb05cc97588R12).

File: dts/bindings/gpio/nordic,nrf-gpiote.yaml
Comment (anangl) [inline dts/bindings/gpio/nordic,nrf-gpiote.yaml]: NRF5 -> nRF
Same applies to [line 12](https://github.com/zephyrproject-rtos/zephyr/pull/8729/files#diff-421ee7681169f4fe7e16f5bada858df5R12).

**Comment by Mierunski** [comment]
@galak @ioannisg @carlescufi Could we get this merged?

**Comment by carlescufi** [comment]
@Mierunski yep, I just dismissed @galak 's review since it was addressed. @ioannisg can you please revisit this so we can merge it?

File: arch/arm/soc/nordic_nrf/nrf52/dts.fixup
Comment (ioannisg) [inline arch/arm/soc/nordic_nrf/nrf52/dts.fixup]: nRF52832 does not have P1, AFAICS. Would it make sense to skip this define for that SOC, and leave it for 52840 only?

**Review by ioannisg — CHANGES_REQUESTED** [review]
Just one minor issue to address, otherwise this looks ok to me.

File: arch/arm/soc/nordic_nrf/nrf52/dts.fixup
Comment (Mierunski) [reply arch/arm/soc/nordic_nrf/nrf52/dts.fixup]: It would require additional fixup file, and in current way it doesn't affect nRF52832 so I would keep it that way, especially that we are gonna get rid of it as soon as code generation gets merged.

Comment (ioannisg) [reply arch/arm/soc/nordic_nrf/nrf52/dts.fixup]: I understand that it does not affect nRF53832, I am wondering, now, why it would need a separate fixup? Could we just add #ifdef inside the fixup for nRF53840?

Comment (Mierunski) [reply arch/arm/soc/nordic_nrf/nrf52/dts.fixup]: Ok, I've added required if

**Review by ioannisg — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
recheck

File: drivers/gpio/gpio_nrfx.c
Comment (pdunaj) [inline drivers/gpio/gpio_nrfx.c]: There is a bug here!
gpiote_channel_alloc returns the channel number so if more then one pin will enable edge trigger it will fail as channel will be non-zero!

Comment (pdunaj) [reply drivers/gpio/gpio_nrfx.c]: This must be fixed asap

Comment (pdunaj) [reply drivers/gpio/gpio_nrfx.c]: @carlescufi @Mierunski FYI

Comment (pdunaj) [reply drivers/gpio/gpio_nrfx.c]: besides it seems that when I disable callbacks they are still coming!
how was it tested?

Comment (pdunaj) [reply drivers/gpio/gpio_nrfx.c]: I disable the callback in the interrupt but I get another interrupt just after that....
This was working fine before this change

Comment (pdunaj) [reply drivers/gpio/gpio_nrfx.c]: https://github.com/zephyrproject-rtos/zephyr/issues/9535

**Comment by carlescufi** [comment]
Thanks! Will assign to Mieszko, it’s his code

**Comment by pdunaj** [comment]
@carlescufi Thanks, I think @Mierunski is off today. I fixed two bugs there, should push the change in few minutes.

**Pull request merged**
